src/repository.rs

```diff
@@ -117,136 +125,155 @@ impl<D> Repository<D> for FileSystemRepository<D>
+    fn fetch_target<'a>(
+        &'a self,
+        target_path: &'a TargetPath,
+        target_description: &'a TargetDescription,
         min_bytes_per_second: u32,
-    ) -> Result<SafeReader<Self::TargetRead>> {
-        let mut path = self.local_path.join("targets");
-        path.extend(target_path.components());
+    ) -> TufFuture<'a, Result<Box<dyn AsyncRead>>> {
+        Box::pinned(
+            async move {
+                let mut path = self.local_path.join("targets");
+                path.extend(target_path.components());
 
-        if !path.exists() {
-            return Err(Error::NotFound);
-        }
+                if !path.exists() {
+                    return Err(Error::NotFound);
+                }
 
-        let (alg, value) = crypto::hash_preference(target_description.hashes())?;
+                let (alg, value) = crypto::hash_preference(target_description.hashes())?;
+
+                let reader: Box<dyn AsyncRead> = Box::new(SafeReader::new(
+                    AllowStdIo::new(File::open(&path)?),
+                    target_description.size(),
+                    min_bytes_per_second,
+                    Some((alg, value.clone())),
+                )?);
 
-        SafeReader::new(
-            File::open(&path)?,
-            target_description.size(),
-            min_bytes_per_second,
-            Some((alg, value.clone())),
+                Ok(reader)
+            },
         )
     }
 }
 
-fn atomically_write<F>(path: &Path, mut f: F) -> Result<()>
-where
-    F: FnMut(&mut Write) -> Result<()>,
-{
-    // We want to atomically write the file to make sure clients can never see a partially written file.
-    // In order to do this, we'll write to a temporary file in the same directory as our target, otherwise
-    // we risk writing the temporary file to one mountpoint, and then non-atomically copying the file to another mountpoint.
+fn create_temp_file(path: &Path) -> Result<NamedTempFile> {
+    // We want to atomically write the file to make sure clients can never see a partially written
+    // file.  In order to do this, we'll write to a temporary file in the same directory as our
+    // target, otherwise we risk writing the temporary file to one mountpoint, and then
+    // non-atomically copying the file to another mountpoint.
 
-    let mut temp_file = if let Some(parent) = path.parent() {
+    if let Some(parent) = path.parent() {
         DirBuilder::new().recursive(true).create(parent)?;
-        NamedTempFile::new_in(parent)?
+        Ok(NamedTempFile::new_in(parent)?)
     } else {
-        NamedTempFile::new_in(".")?
-    };
-
-    f(&mut temp_file)?;
-
-    temp_file.persist(&path)?;
-
-    Ok(())
+        Ok(NamedTempFile::new_in(".")?)
+    }
 }
 
 /// A repository accessible over HTTP.
-pub struct HttpRepository<D>
+pub struct HttpRepository<C, D>
 where
+    C: Connect + Sync,
     D: DataInterchange,
 {
     url: Url,
-    client: Client,
+    client: Client<C>,
```

Comment:
What was your reasoning for moving the client into the repository? My original API design was to have the client be the primary entity a user interacts with.